ring/spread_minimizing_token_generator.go

```diff
@@ -0,0 +1,279 @@
+
+import (
+	"container/heap"
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"golang.org/x/exp/slices"
+)
+
+const totalTokensCount = 1 << 32
+
+var (
+	instanceIDRegex          = regexp.MustCompile(`^(.*)-(zone-[a-z])-(\d+)$`)
+	errorBadInstanceIDFormat = func(instanceID string) error {
+		return fmt.Errorf("unable to extract instance id from \"%s\"", instanceID)
+	}
+	errorZoneNotValid = func(zone string) error {
+		return fmt.Errorf("zone %s is not valid", zone)
+	}
+	errorMultipleOfZonesCount = func(zonesCount, optimalTokenOwnership uint32, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: optimal token ownership must be a positive multiple of number of zones %d", token.prevToken, token.token, optimalTokenOwnership, zonesCount)
+	}
+	errorLowerAndUpperBoundModulo = func(zonesCount, optimalTokenOwnership uint32, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: lower and upper bounds must be congruent modulo number of zones %d", token.prevToken, token.token, optimalTokenOwnership, zonesCount)
+	}
+	errorDistanceBetweenTokensNotBigEnough = func(optimalTokenOwnership, ownership int, token ringToken) error {
+		return fmt.Errorf("calculation of a new token between %d and %d with optimal token ownership %d was impossible: distance between lower and upper bound %d is not big enough", token.prevToken, token.token, optimalTokenOwnership, ownership)
+	}
+)
+
+type SpreadMinimizingConfig struct {
+	InstanceID        string
+	Zone              string
+	TokensPerInstance int
+}
+
+type SpreadMinimizingTokenGenerator struct {
+	cfg        SpreadMinimizingConfig
+	instanceID int
+	zoneID     int
+	zones      []string
+	logger     log.Logger
+}
+
+func NewSpreadMinimizingTokenGenerator(cfg SpreadMinimizingConfig, zones []string, logger log.Logger) (*SpreadMinimizingTokenGenerator, error) {
+	sortedZones := make([]string, len(zones))
+	copy(sortedZones, zones)
+	if !slices.IsSorted(sortedZones) {
+		sort.Strings(sortedZones)
+	}
+	instanceID, err := getInstanceID(cfg.InstanceID)
+	if err != nil {
+		return nil, err
+	}
+	zoneID, err := getZoneID(cfg.Zone, sortedZones)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenGenerator := &SpreadMinimizingTokenGenerator{
+		cfg:        cfg,
+		instanceID: instanceID,
+		zoneID:     zoneID,
+		zones:      sortedZones,
+		logger:     logger,
+	}
+	return tokenGenerator, nil
+}
+
+func getInstanceID(instanceID string) (int, error) {
+	parts := instanceIDRegex.FindStringSubmatch(instanceID)
+	if len(parts) != 4 {
+		return 0, errorBadInstanceIDFormat(instanceID)
+	}
+	return strconv.Atoi(parts[3])
+}
+
+// getZoneID gets a zone name and a slice of sorted zones,
+// and return the index of the zone in the slice.
+func getZoneID(zone string, sortedZones []string) (int, error) {
```

Comment:
Nit, and slightly optionated, but there aren't really _getters_ (and as mentioned in previous comment, getters shouldn't have a `get` prefix), so I'd call these methods:

```go 
func parseInstanceID(instanceID string) (int, error)

func findZoneID(zone string, sortedZones []string) (int, error)
```

Because the first one does the parsing of the string, while the second one tries to find it in the list of provided zones.